### ISSUE TITLE
Update: Failed to Unregister old appimage and register new one in menu

### ISSUE THREAD

**Issue by star-buck** [TheAssassin/AppImageLauncher#56]
I installed this appimage from appimagehub.com and then found an update via right-click check for updates in taskmanager.
After update went fine, it threw this error:
![screenshot_20180701_195715](https://user-images.githubusercontent.com/840070/42137340-18446f20-7d6b-11e8-896c-e0545cde9b3a.png)

**Comment by star-buck** [comment]
note: i now realized i tried this while the installed appimage was still running, so lilely this was the cause of failure :) Still we need to check for that and gracefully find a solution?

**Comment by TheAssassin** [comment]
@star-buck great find, I've faced a similar issue on my dev system maybe a week ago, and now I realize it's been the same situation.

It seems like you're using an outdated version, since I can't find this message in the translations any more. See https://translate.assassinate-you.net/translate/appimagelauncher/ui/en/?q=failed+to+unregister&search=ftx&source=1&target=1&type=all&offset=0.

IIRC, the message was shown when AppImageLauncher failed to remove the file during the unintegration process. The error message isn't really telling much about the issue. I've since improved this a bit. The new error message should say "Failed to remove AppImage: <error message>".

If I'd remove a normal binary, it'd just disappear from the visible filesystem, and the blocks would silently be freed when the application exits. In our case, we have a FUSE filesystem mounted, and I suppose the system would rather report an error instead of removing the file when the file gets unmounted.

**Comment by star-buck** [comment]
One other idea would be to check if appimage is running and then do not
allow update, but show a notice: please close appimage before updating
(check systray too!).

**Comment by TheAssassin** [comment]
We could perhaps combine this with the other method I suggested, but we shouldn't solely rely on it. A user could still open the app during the update, and we can't prevent them from doing so (unless we introduce some kind of locking mechanism, since AppImageLauncher is always triggeren on launch it could show an error message then).

**Comment by ferdnyc** [comment]
Another option would be to make use of the GIO `trash:///` folder, as I just accidentally demonstrated to myself that Linux/fuse have _absolutely no issues_ with moving a mounted filesystem image into the trash.

```console
$ find ~/ -name OpenShot-v2.4.2-x86_64.AppImage
/home/ferd/.local/share/Trash/files/OpenShot-v2.4.2-x86_64.AppImage
$ mount|grep AppImage
OpenShot-v2.4.2-x86_64.AppImage on /tmp/.mount_a8DUdU type fuse.OpenShot-v2.4.2-x86_64.AppImage (ro,nosuid,nodev,relatime,user_id=1000,group_id=1000)
```

**Comment by ferdnyc** [comment]
(The AppImage still unmounted normally at application exit, in the `.AppImage`-file-in-the-`trash:///` scenario.)

**Comment by TheAssassin** [comment]
Well, I'd rather try not to depend on any desktop environment specific tools. I'm not sure whether this feature is available on e.g., KDE, or even on some less popular DEs like i3.

But the idea behind your suggestion is really good, actually. You're right, moving a mounted file is easier than removing the file, since moving a file does only change some metadata, and apparently FUSE doesn't have a problem with it. We could just implement our own "AppImage trash" in `~/Applications`, which is emptied from time to time using the methods suggested above. This improves the UX a lot, since the file will at least disappear from `~/Applications` on removal.

**Comment by star-buck** [comment]
Sounds like a cool idea.

**Comment by TheAssassin** [comment]
Working on the implementation of the combination of these suggestions.

**Comment by TheAssassin** [comment]
@ferdnyc actually I missed one point about your suggestion: it is about removing AppImage, but not about updating AppImages.

When performing an update of an AppImage, AppImageUpdate leaves the old AppImage in place, and will normally create a new file with the new AppImage's filename. Only if both the new and old file have the same name, AppImageUpdate will touch the file. But it doesn't delete it but renames it (appends `.zs-old` to the file's name). AppImageLauncher's update tool does exactly that, but in comparison to AppImageUpdate it automatically integrates the updated file. After the update, a user should have both old and new version available.

@star-buck's problem was about "unregistering" the AppImage. This process just removes the icons and the desktop file that have been copied into place when first integrating the AppImage. It only reads the AppImage but doesn't change the file. I guess we need @star-buck to update and perhaps try to reproduce the issue. I am trying to reproduce it as well. I still think @star-buck is using an older version of AppImageLauncher, whose error message is wrong, and therefore we might get a wrong impression of what led to this error.

Nevertheless, I will continue to develop the trash bin idea, because I really like it, and it solves issues about the removal of the AppImage. As there's the possibility to remove the AppImage right away, the trash bin will also improve the situation there.

**Comment by TheAssassin** [comment]
On my system (btrfs filesystem) removing a running AppImage doesn't hurt. I just updated a copy of AppImageUpdate using the update helper, and checked the box asking me whether to remove the file. This worked without issues. But as there's so many different possible configurations, this might not always be true.

I think the proper solution is to continue with the trash bin idea, i.e., before removing AppImages, move them into the bin, then call a cleanup method that tries to remove the files in there, but will just silently skip the ones for which the OS reports they can't be removed. The cleanup method is run by every AppImageLauncher tool, so by using AppImageLauncher, the directory is cleaned up regularly without any user interaction. This is how desktop files are cleaned up as well. I think that's "good enough" until the daemon has been implemented which could decrease the cleanup intervals and therefore reduce the time until AppImages are removed from the system as soon as this is possible.

Also, when AppImages are moved into the trash bin, they're made non-executable. This is not really effective in preventing people from using those files, but more of a cosmetic change.
